experimental/Schemes/Resolution_structure.jl

```diff
@@ -1,54 +1,138 @@
-#    n = length(maps)
-#    for i in 1:n-1
-#      @assert domain(maps[i]) === codomain(maps[i+1]) "not a sequence of morphisms"
-#    end
-#    return new{typeof(domain(maps[end])),typeof(codomain(first(maps)))}(maps)
-#  end
-#end
+@attributes mutable struct MixedBlowUpSequence{
+                                                DomainType<:AbsCoveredScheme,
+                                                CodomainType<:AbsCoveredScheme
+                                              }<:AbsDesingMor{  DomainType,
+                                                                CodomainType, 
+                                                                MixedBlowUpSequence{DomainType, CodomainType}
+                                              }
+  maps::Vector{Union{<:BlowupMorphism,<:NormalizationMorphism}}       # count right to left:
+                                                 # original scheme is codomain of map 1
+  # boolean flags
+  resolves_sing::Bool                            # domain not smooth yet?
+  is_trivial::Bool                               # codomain already smooth?
+
+  # fields for caching, to be filled during desingularization
+  ex_div::Vector{AbsIdealSheaf}                  # list of exc. divisors arising from individual steps
+                                                 # in domain(maps[end]) -- at least in the end
+  normal_steps::Vector{Int}                      # index of parts of preimage of singular locus 
+                                                 # not arising from blow-ups
+  ## CAUTION: ex_div[normal_steps[i]] need not even be of pure codim 1, i.e. a divisor, 
+  ##                                  as long as resolves_sing == false
+  ##          lower dimensional components of it belong to future centers of blow ups 
+
+
+  # fields for caching to be filled a posteriori (on demand, only if partial_res==false)
+  underlying_morphism::CompositeCoveredSchemeMorphism{DomainType, CodomainType}
+  exceptional_divisor::WeilDivisor          
+
+  function MixedBlowUpSequence(maps::Vector{<:AbsCoveredSchemeMorphism})
+    n = length(maps)
+    for i in 1:n
+      @assert all(x->((x isa BlowupMorphism) || (x isa NormalizationMorphism)), maps) "only blow-ups and normalizations allowed"
+    end
+    for i in 1:n-1
+      @assert domain(maps[i]) === codomain(maps[i+1]) "not a sequence of morphisms"
+    end
+    resi = new{typeof(domain(maps[end])),typeof(codomain(first(maps)))}(maps)
+    resi.normal_steps = [i for i in 1:n if maps[i] isa NormalizationMorphism]
+    return resi
+  end
+
+end
 
 
 ##################################################################################################
 # getters
 ##################################################################################################
+underlying_morphism(phi::NormalizationMorphism) = CoveredSchemeMorphism(phi.X,phi.Y,phi.f)
+domain(phi::NormalizationMorphism) = phi.X
+codomain(phi::NormalizationMorphism) = phi.Y
```

Comment:
these getters will work automatically once `underlying_morphism` is implemented as sketched above.

Reply:
done